File: lib/features/analytics/domain/insights.dart
import 'analytics_view.dart';
import 'goal_scoring.dart';

/// What an insight is about. The UI turns this plus [Insight.values] into
/// copy -- the rules themselves stay free of strings so the same rule can
/// speak English or Hebrew, and so a rule is testable on its numbers alone.
enum InsightKind {
  plateau,
  personalBest,
  proteinShortfall,
  calorieDrift,
  volumeDrop,
  sleepDebt,
  consistencyWin,
  neglectedMuscle,
}

enum InsightTone { positive, neutral, warning }

/// One generated observation.
class Insight {
  final InsightKind kind;
  final InsightTone tone;

  /// Exercise id or muscle name, for the kinds that name one.
  final String? subjectId;

  /// The numbers the copy interpolates. Keys are per-kind and documented on
  /// the rule that emits them.
  final Map<String, double> values;

  /// Higher wins when more insights qualify than the card will show.
  final int priority;

  const Insight({
    required this.kind,
    required this.tone,
    required this.priority,
    this.subjectId,
    this.values = const {},
  });
}

/// How many insights the card shows. Past three it stops being a summary and
/// becomes a second screen the user has to read.
const int maxInsights = 3;

/// Runs every rule and returns the strongest [maxInsights].
///
/// Each rule is a pure function of the already-computed view, so adding one is
/// a function and a unit test -- no query, no UI change.
List<Insight> generateInsights(AnalyticsView view) {
  final found = <Insight>[
    ..._plateaus(view),
    ..._personalBests(view),
    ...<Insight?>[
      _proteinShortfall(view),
      _calorieDrift(view),
      _volumeDrop(view),
      _sleepDebt(view),
      _consistencyWin(view),
      _neglectedMuscle(view),
    ].whereType<Insight>(),
  ]..sort((a, b) => b.priority.compareTo(a.priority));

  return found.take(maxInsights).toList();
}

/// values: `weight` (kg), `days`, `sessions`.
Iterable<Insight> _plateaus(AnalyticsView view) =>
    view.plateaus.where((p) => p.isPlateau).take(2).map((p) => Insight(
          kind: InsightKind.plateau,
          tone: InsightTone.warning,
          subjectId: p.exerciseId,
          // The longer the stall, the more it deserves the slot.
          priority: 60 + p.daysSinceIncrease,
          values: {
            'weight': p.lastTopWeightKg,
            'days': p.daysSinceIncrease.toDouble(),
            'sessions': p.sessionsAtWeight.toDouble(),
          },
        ));

/// values: `e1rm` (kg).
Iterable<Insight> _personalBests(AnalyticsView view) =>
    view.plateaus.where((p) => p.isPersonalBest).take(1).map((p) => Insight(
          kind: InsightKind.personalBest,
          tone: InsightTone.positive,
          subjectId: p.exerciseId,
          priority: 90,
          values: {
            'e1rm': view.strength[p.exerciseId]?.sessions.last.e1rm ?? 0,
          },
        ));

/// values: `actual` (g), `goal` (g).
Insight? _proteinShortfall(AnalyticsView view) {
  final goal = view.targets.proteinGoal;
  if (goal == null || goal <= 0) return null;
  final recent = view.protein.tailAverage(7);
  // Two observations minimum: one low day is a day, not a trend.
  if (recent == null || view.protein.observedCount < 2) return null;
  if (recent >= goal * 0.85) return null;

  return Insight(
    kind: InsightKind.proteinShortfall,
    tone: InsightTone.warning,
    priority: 70,
    values: {'actual': recent, 'goal': goal},
  );
}

/// values: `actual` (kcal), `goal` (kcal), `direction` (-1 under, 1 over).
Insight? _calorieDrift(AnalyticsView view) {
  final goal = view.targets.calorieGoal;
  if (goal == null || goal <= 0) return null;
  final recent = view.calories.tailAverage(7);
  if (recent == null || view.calories.observedCount < 3) return null;

  final ratio = recent / goal;
  if (ratio > 0.85 && ratio < 1.15) return null;

  return Insight(
    kind: InsightKind.calorieDrift,
    tone: InsightTone.neutral,
    priority: 50,
    values: {
      'actual': recent,
      'goal': goal,
      'direction': ratio >= 1 ? 1 : -1,
    },
  );
}

/// values: `drop` (0..1 share below the earlier average).
///
/// Compares the latest bucket against the average of the ones before it, which
/// is why it needs at least three: with two, "the average of the rest" is a
/// single bucket and any normal week-to-week swing trips it.
Insight? _volumeDrop(AnalyticsView view) {
  final points = view.volume.points.where((p) => p.hasValue).toList();
  if (points.length < 3) return null;

  final latest = points.last.value!;
  final earlier = points.sublist(0, points.length - 1).map((p) => p.value!);
  final baseline = earlier.reduce((a, b) => a + b) / earlier.length;
  if (baseline <= 0 || latest >= baseline * 0.7) return null;

  return Insight(
    kind: InsightKind.volumeDrop,
    tone: InsightTone.warning,
    priority: 65,
    values: {'drop': 1 - (latest / baseline)},
  );
}

/// values: `nights`, `goal` (hours).
Insight? _sleepDebt(AnalyticsView view) {
  final recent = view.goalDays.reversed.take(7).toList();
  if (recent.length < 7) return null;

  final short = recent.where((d) => !d.met.contains(WellnessGoal.sleep)).length;
  if (short < 3) return null;

  return Insight(
    kind: InsightKind.sleepDebt,
    tone: InsightTone.warning,
    priority: 55,
    values: {
      'nights': short.toDouble(),
      'goal': view.targets.sleepGoalHours,
    },
  );
}

/// values: `days`.
Insight? _consistencyWin(AnalyticsView view) {
  if (view.currentStreak < 7) return null;
  return Insight(
    kind: InsightKind.consistencyWin,
    tone: InsightTone.positive,
    priority: 85,
    values: {'days': view.currentStreak.toDouble()},
  );
}

/// The muscle group with the fewest sets, when the user is training enough for
/// its absence to be a choice rather than a coincidence.
///
/// values: `sets`.
Insight? _neglectedMuscle(AnalyticsView view) {
  final total = view.setsPerMuscle.values.fold(0, (a, b) => a + b);
  // Under ~30 sets in range there is not enough training to call anything
  // neglected -- everything looks neglected in a light week.
  if (total < 30 || view.setsPerMuscle.length < 3) return null;

  final sorted = view.setsPerMuscle.entries.toList()
    ..sort((a, b) => a.value.compareTo(b.value));
  final lowest = sorted.first;
  if (lowest.value > total * 0.05) return null;

  return Insight(
    kind: InsightKind.neglectedMuscle,
    tone: InsightTone.neutral,
    subjectId: lowest.key,
    priority: 40,
    values: {'sets': lowest.value.toDouble()},
  );
}

File: Packages/WellnessKit/Sources/WellnessDomain/Insights.swift
import Foundation
import WellnessModels

public let maxInsights = 3

public func generateInsights(_ view: AnalyticsView) -> [Insight] {
    var found: [Insight] = []
    found.append(contentsOf: plateauInsights(view))
    found.append(contentsOf: personalBestInsights(view))
    if let i = proteinShortfall(view) { found.append(i) }
    if let i = calorieDrift(view) { found.append(i) }
    if let i = volumeDrop(view) { found.append(i) }
    if let i = sleepDebt(view) { found.append(i) }
    if let i = consistencyWin(view) { found.append(i) }
    if let i = neglectedMuscle(view) { found.append(i) }
    found.sort { $0.priority > $1.priority }
    return Array(found.prefix(maxInsights))
}

private func plateauInsights(_ view: AnalyticsView) -> [Insight] {
    view.plateaus.filter(\.isPlateau).prefix(2).map { p in
        Insight(
            kind: .plateau, tone: .warning, subjectId: p.exerciseId,
            values: [
                "weight": p.lastTopWeightKg,
                "days": Double(p.daysSinceIncrease),
                "sessions": Double(p.sessionsAtWeight),
            ],
            priority: 60 + p.daysSinceIncrease
        )
    }
}

private func personalBestInsights(_ view: AnalyticsView) -> [Insight] {
    view.plateaus.filter(\.isPersonalBest).prefix(1).map { p in
        Insight(
            kind: .personalBest, tone: .positive, subjectId: p.exerciseId,
            values: [
                "e1rm": view.strength[p.exerciseId]?.sessions.last?.e1rm ?? 0
            ],
            priority: 90
        )
    }
}

private func proteinShortfall(_ view: AnalyticsView) -> Insight? {
    guard let goal = view.targets.proteinGoal, goal > 0 else { return nil }
    guard let recent = view.protein.tailAverage(7), view.protein.observedCount >= 2 else { return nil }
    guard recent < goal * 0.85 else { return nil }
    return Insight(kind: .proteinShortfall, tone: .warning,
                   values: ["actual": recent, "goal": goal], priority: 70)
}

private func calorieDrift(_ view: AnalyticsView) -> Insight? {
    guard let goal = view.targets.calorieGoal, goal > 0 else { return nil }
    guard let recent = view.calories.tailAverage(7), view.calories.observedCount >= 3 else { return nil }
    let ratio = recent / goal
    guard ratio <= 0.85 || ratio >= 1.15 else { return nil }
    return Insight(kind: .calorieDrift, tone: .neutral,
                   values: ["actual": recent, "goal": goal, "direction": ratio >= 1 ? 1 : -1],
                   priority: 50)
}

private func volumeDrop(_ view: AnalyticsView) -> Insight? {
    let points = view.volume.points.filter(\.hasValue)
    guard points.count >= 3, let latest = points.last?.value else { return nil }
    let earlier = points.dropLast().compactMap(\.value)
    let baseline = earlier.reduce(0, +) / Double(earlier.count)
    guard baseline > 0, latest < baseline * 0.7 else { return nil }
    return Insight(kind: .volumeDrop, tone: .warning,
                   values: ["drop": 1 - (latest / baseline)], priority: 65)
}

private func sleepDebt(_ view: AnalyticsView) -> Insight? {
    let recent = Array(view.goalDays.suffix(7))
    guard recent.count >= 7 else { return nil }
    let short = recent.filter { !$0.met.contains(.sleep) }.count
    guard short >= 3 else { return nil }
    return Insight(kind: .sleepDebt, tone: .warning,
                   values: ["nights": Double(short), "goal": view.targets.sleepGoalHours],
                   priority: 55)
}

private func consistencyWin(_ view: AnalyticsView) -> Insight? {
    guard view.currentStreak >= 7 else { return nil }
    return Insight(kind: .consistencyWin, tone: .positive,
                   values: ["days": Double(view.currentStreak)], priority: 85)
}

private func neglectedMuscle(_ view: AnalyticsView) -> Insight? {
    let total = view.setsPerMuscle.values.reduce(0, +)
    guard total >= 30, view.setsPerMuscle.count >= 3 else { return nil }
    let sorted = view.setsPerMuscle.sorted { $0.value < $1.value }
    guard let lowest = sorted.first, lowest.value <= total / 20 else { return nil }
    return Insight(kind: .neglectedMuscle, tone: .neutral, subjectId: lowest.key,
                   values: ["sets": Double(lowest.value)], priority: 40)
}
